File: everything_parsing/06-ambiguity-disambiguation/java/Ambiguity.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Module 06 — find every parse tree a grammar gives a string.
 *
 * A grammar is AMBIGUOUS if some string has two or more parse trees. Because a
 * parse tree is in bijection with its leftmost derivation (Module 5), it is
 * enough to enumerate every leftmost derivation that yields the target string.
 *
 * The search is a breadth-first walk over sentential forms. Two prunes keep it
 * finite and fast:
 *   - terminals never disappear, so drop any form with more terminals than |w|
 *   - in a leftmost derivation everything left of the leftmost nonterminal is
 *     final, so that terminal prefix must already match the target
 */
public final class Ambiguity {
    private Ambiguity() {}

    private static final int MAX_STATES = 1_000_000;

    private record State(List<String> form, int[] choices) {}

    /** Every leftmost-derivation choice sequence that yields `target`. */
    static List<int[]> allLeftmostDerivations(Grammar g, List<String> target) {
        List<int[]> out = new ArrayList<>();
        Deque<State> queue = new ArrayDeque<>();
        queue.add(new State(List.of(g.start), new int[0]));

        int budget = MAX_STATES;
        while (!queue.isEmpty() && budget-- > 0) {
            State st = queue.poll();
            int i = Derivation.leftmostNonterminal(g, st.form);
            if (i < 0) {
                if (st.form.equals(target)) out.add(st.choices);
                continue;
            }
            if (i > target.size()) continue;
            if (!st.form.subList(0, i).equals(target.subList(0, i))) continue; // prefix must match

            for (Grammar.Production p : g.productionsFor(st.form.get(i))) {
                List<String> next = splice(st.form, i, p.rhs());
                if (terminalCount(g, next) > target.size()) continue;
                queue.add(new State(next, append(st.choices, p.index())));
            }
        }
        return out;
    }

    /** Distinct parse trees for `target`, keyed by their rendered form. */
    static List<ParseTree> allTrees(Grammar g, List<String> target) {
        Map<String, ParseTree> distinct = new LinkedHashMap<>();
        for (int[] choices : allLeftmostDerivations(g, target)) {
            ParseTree t = ParseTree.build(g, choices, /*leftmost*/ true);
            distinct.putIfAbsent(t.render(), t);
        }
        return new ArrayList<>(distinct.values());
    }

    static boolean isAmbiguousFor(Grammar g, List<String> target) {
        return allTrees(g, target).size() > 1;
    }

    /** Smallest string (by length, then lexicographic) with 2+ trees; null if none up to maxLen. */
    static List<String> smallestAmbiguousString(Grammar g, int maxLen) {
        for (String w : Derivation.enumerate(g, maxLen)) {
            List<String> toks = w.equals("epsilon") ? List.of() : List.of(w.split(" "));
            if (allTrees(g, toks).size() > 1) return toks;
        }
        return null;
    }

    // ── helpers ──

    private static List<String> splice(List<String> form, int i, List<String> rhs) {
        List<String> out = new ArrayList<>(form.size() + rhs.size());
        out.addAll(form.subList(0, i));
        out.addAll(rhs);
        out.addAll(form.subList(i + 1, form.size()));
        return out;
    }
    private static int terminalCount(Grammar g, List<String> form) {
        int n = 0;
        for (String s : form) if (!g.isNonterminal(s)) n++;
        return n;
    }
    private static int[] append(int[] a, int x) {
        int[] b = new int[a.length + 1];
        System.arraycopy(a, 0, b, 0, a.length);
        b[a.length] = x;
        return b;
    }
}

File: everything_parsing/06-ambiguity-disambiguation/js/ambiguity.mjs
// Module 06 — find every parse tree a grammar gives a string.
//
// A grammar is AMBIGUOUS if some string has two or more parse trees. Since a
// parse tree is in bijection with its leftmost derivation (Module 5), it is
// enough to enumerate every leftmost derivation that yields the target.
//
// Breadth-first over sentential forms, with two prunes:
//   - terminals never disappear → drop forms with more terminals than |w|
//   - in a leftmost derivation everything left of the leftmost nonterminal is
//     final → that terminal prefix must already match the target

import { leftmostNonterminal, enumerate } from "./derivation.mjs";
import { ParseTree } from "./parsetree.mjs";

const MAX_STATES = 1_000_000;

export function allLeftmostDerivations(g, target) {
  const out = [];
  const queue = [{ form: [g.start], choices: [] }];
  let budget = MAX_STATES;

  while (queue.length > 0 && budget-- > 0) {
    const st = queue.shift();
    const i = leftmostNonterminal(g, st.form);
    if (i < 0) {
      if (sameList(st.form, target)) out.push(st.choices);
      continue;
    }
    if (i > target.length) continue;
    if (!sameList(st.form.slice(0, i), target.slice(0, i))) continue;

    for (const p of g.productionsFor(st.form[i])) {
      const next = [...st.form.slice(0, i), ...p.rhs, ...st.form.slice(i + 1)];
      if (terminalCount(g, next) > target.length) continue;
      queue.push({ form: next, choices: [...st.choices, p.index] });
    }
  }
  return out;
}

export function allTrees(g, target) {
  const distinct = new Map();
  for (const choices of allLeftmostDerivations(g, target)) {
    const t = ParseTree.build(g, choices, true);
    const key = t.render();
    if (!distinct.has(key)) distinct.set(key, t);
  }
  return [...distinct.values()];
}

export function isAmbiguousFor(g, target) {
  return allTrees(g, target).length > 1;
}

export function smallestAmbiguousString(g, maxLen) {
  for (const w of enumerate(g, maxLen)) {
    const toks = w === "epsilon" ? [] : w.split(" ");
    if (allTrees(g, toks).length > 1) return toks;
  }
  return null;
}

// ── helpers ──

function terminalCount(g, form) {
  let n = 0;
  for (const s of form) if (!g.isNonterminal(s)) n++;
  return n;
}
function sameList(a, b) { return a.length === b.length && a.every((x, i) => x === b[i]); }
